### ISSUE TITLE
RFC: Use src/ directory

### ISSUE THREAD

**Issue by sloria** [marshmallow-code/marshmallow#1017]
Put the package in `src/marshmallow`, as recommended in these articles:

* https://blog.ionelmc.ro/2014/05/25/python-packaging/#the-structure
* https://hynek.me/articles/testing-packaging/

The benefits makes sense, but I haven't yet run into the pitfalls talked about in those articles...

Strong opinions welcome!

**Comment by lafrech** [comment]
I read those posts one or two years ago. This totally makes sense.

I tried it and bumped into the coverage issues described in the post and figured the workaround was a bit too complicated for me to bother with the whole thing. So I figured let's courageously wait and see if someone else moves first. I checked a few other projects, and all of them didn't do it, so I left things as they were.

TL; DR I gave it a try and abandoned as it was not straightforward and marshmallow didn't do it.

**Comment by sloria** [comment]
marshmallow shouldn't be dispositive in this regard; I wasn't aware of those blog posts when I started marshmallow. I also checked some other projects at the time, and not many were using `src/`.

Incidentally, @hynek added an update to his article today:

> Quick interlude from 2018: while there is still [disagreement](https://github.com/pypa/python-packaging-user-guide/issues/320) on the src topic, a significant number of projects moved to src since this article has been published. From the four projects above that I used to demonstrate how uncommon it is, two switched to an src layout (Pyramid and Twisted).

**Comment by lafrech** [comment]
> marshmallow shouldn't be dispositive in this regard; I wasn't aware of those blog posts when I started marshmallow

I totally understand that. I just wanted to stress out how much I owe marshmallow about good practices, tests, CI, etc. When in doubt, I always check what is done in marshmallow.

If the coverage issue can be solved/circumvented, let's go for it.

**Comment by sloria** [comment]
Since we aren't using coverage, then it looks like there are no blockers to this.

**Comment by lafrech** [comment]
I'm not sure but it might complicate the merges from 2.x. Not really a blocker as hopefully three won't be many (how long are we maintaining 2.x when 3.0 is out ?) and since a lot of code has changed already, a lot of merges can't be automatic and the changes must be ported manually (I suppose, I can't tell since you're the one performing those tasks).

**Comment by sloria** [comment]
@lafrech If we go forward with this, we should do it on both `2.x-line` and `dev` to avoid complicated merges.
